### DISCUSSION TITLE
Best way to check session still active?

### DISCUSSION THREAD

**Discussion by AJamesPhillips** [supabase/supabase#3409] [Questions]
I'm observing JWT expiration without the auto token refresh seeming to work.  I am experimenting with listening to window.focus event and checking the user is still signed in, otherwise will refresh the whole page.  What's the best way of doing this?

I had thought that doing the following might work:

```javascript
    window_on_focus_listener = () =>
    {
        supabase.auth.update({})
        .then(response =>
        {
            if (response.status === 401)
            {
                refresh_page() // will result in prompt to sign in being displayed
            }
            else console.log("On window focus event, successfully accessed user info.")
        })
        .catch(err =>
        {
            if (err?.status === 401)
            {
                refresh_page() // will result in prompt to sign in being displayed
            }
        })
    }
```

However this does not work as instead the response is:

```javascript
{
    data: null,
    user: null,
    error: Error
```

Where error seems to have a `message` attribute which has a value of "Not logged in.".  I do not want to assume this is a stable error message, i.e. to use `if (response.error?.message === "Not logged in.") refresh_page()`.  What other options / APIs are there?  I'd prefer not to build my own API do perform this check.

**Comment by awalias** [comment]
is the session being persisted to localStorage as per https://github.com/supabase/gotrue-js/blob/4d57cb5db2c833bb08dcb72e1ce025615624efdd/src/GoTrueClient.ts#L654 ?

**Reply by AJamesPhillips** [reply]
I assume it is as I have not changed any defaults and it is able to refresh the session some of the time.

A second related question is how to get the client to auto refresh the session.  From the code it looks like making a new client should trigger the automatic recover / refresh of the session: https://github.com/supabase/gotrue-js/blob/4d57cb5db2c833bb08dcb72e1ce025615624efdd/src/GoTrueClient.ts#L89-L90  however that's insufficient as `_recoverAndRefresh()` is async and not all code paths hook into something that can be listened to like `.auth.onAuthStateChange`.  I'd need to know when `_recoverAndRefresh` has finished so that I can try to check if the session is now working, e.g. by doing a second call to `supabase.auth.update({})`, or which ever API / method is the best  answer to the original question.
